amd_tressfx_vulkan/src/TressFXSimulationVulkan.cpp

```diff
@@ -0,0 +1,1180 @@
+// Shader code generates the per-strand noise from these vectors.
+//
+// This function currently has some hard coded constants, including a hard-coded
+// sin function for magnitude based on frame count.  This will be changed in the next
+// major update, but we maintain it in 3.1 for compatibility.
+//
+//--------------------------------------------------------------------------------------
+void ComputeWindPyramid(float4 &wind, float4 &wind1, float4 &wind2, float4 &wind3,
+                        const float windMag, const tressfx_vec3 &windDir)
+{
+    static int frame = 0;
+
+    float wM = windMag * (pow(sin(frame * 0.05f), 2.0f) + 0.5f);
+
+    tressfx_vec3 windDirN(windDir);
+    windDirN.Normalize();
+
+    tressfx_vec3 XAxis(1.0f, 0, 0);
+    tressfx_vec3 xCrossW = XAxis.Cross(windDirN);
+
+    tressfx_quat rotFromXAxisToWindDir;
+    rotFromXAxisToWindDir.SetIdentity();
+
+    float angle = asin(xCrossW.Length());
+
+    if (angle > 0.001)
+    {
+        rotFromXAxisToWindDir.SetRotation(xCrossW.Normalize(), angle);
+    }
+
+    float angleToWideWindCone = DEG_TO_RAD2(40.f);
+
+    {
+        tressfx_vec3 rotAxis(0, 1.0f, 0);
+
+        tressfx_quat rot(rotAxis, angleToWideWindCone);
+        tressfx_vec3 windVec = rotFromXAxisToWindDir * rot * XAxis * wM;
+        wind = float4(windVec.x, windVec.y, windVec.z, (float)frame);
+    }
+
+    {
+        tressfx_vec3 rotAxis(0, -1.0f, 0);
+        tressfx_quat rot(rotAxis, angleToWideWindCone);
+        tressfx_vec3 windVec = rotFromXAxisToWindDir * rot * XAxis * wM;
+        wind1 = float4(windVec.x, windVec.y, windVec.z, (float)frame);
+    }
+
+    {
+        tressfx_vec3 rotAxis(0, 0, 1.0f);
+        tressfx_quat rot(rotAxis, angleToWideWindCone);
+        tressfx_vec3 windVec = rotFromXAxisToWindDir * rot * XAxis * wM;
+        wind2 = float4(windVec.x, windVec.y, windVec.z, (float)frame);
+    }
+
+    {
+        tressfx_vec3 rotAxis(0, 0, -1.0f);
+        tressfx_quat rot(rotAxis, angleToWideWindCone);
+        tressfx_vec3 windVec = rotFromXAxisToWindDir * rot * XAxis * wM;
+        wind3 = float4(windVec.x, windVec.y, windVec.z, (float)frame);
+    }
+
+    frame++;
+}
+
+// scale the stiffness value based on current and target frames rates and minimum
+// stiffness value.
+float getScaledStiffness(float s0, float s_min_scale, float h, float h0)
+{
+    float s_min = s0 * s_min_scale;
+    float s = ((s0 - s_min) / h0) * h + s_min;
+    return s;
+}
+
+//--------------------------------------------------------------------------------------
+//
+// Simulate
+//
+// Runs the hair simulation which will animate the hair vertices. This function calls
+// Dispatch to execute compute shaders that implement the simulation.
+//
+////--------------------------------------------------------------------------------------
+VkResult TressFXSimulation::Simulate(VkDevice pvkDevice, VkCommandBuffer commandBuffer,
```

Comment:
This *huge* function could benefit a lot from splitting it up into a couple separate pieces.